functions/src/scripts/converter.ts

```diff
@@ -0,0 +1,336 @@
+            const tempVersion = CVETree[ID]['BulletinVersion'];
+            delete CVETree[ID]['BulletinVersion'];
+            const tempLongVersionNumber = "BulletinVersion:" + tempVersion;
+            ref.child(ID).child(tempLongVersionNumber).set(CVETree[ID]).catch(error => { console.log(error) });
+        }
+    } else {
+        for (const CVE in CVETree) {
+            if (!Object.values(CVETree).includes(CVE)) {
+                const tempVersion = CVETree[CVE]['BulletinVersion'];
+                delete CVETree[CVE]['BulletinVersion'];
+                const tempLongVersionNumber = "BulletinVersion:" + tempVersion;
+                ref.child(CVE).child(tempLongVersionNumber).set(CVETree[CVE]).catch(error => { console.log(error) });
+            }
+        }
+        for (const CVE in tree) {
+            //check if version in history tree
+            const currentTreeCVEData = tree[CVE];
+            let hasKey = false;
+            if (JSONData[CVE] !== undefined) {
+                for (const key of Object.keys(currentTreeCVEData)) {
+                    if (key === longVersionNumberHistory) {
+                        hasKey = true;
+                    }
+                }
+                if (!hasKey) {
+                    //if doesn't have version add in the data
+                    const cveData: Record<string, object> = {};
+                    cveData['published_date'] = JSONData.published;
+                    for (const key of Object.keys(JSONData[CVE])) {
+                        cveData[key] = JSONData[CVE][key];
+                    }
+                    const constructedData = JSON.parse(JSON.stringify(cveData));
+                    delete constructedData['BulletinVersion'];
+                    ref.child(CVE).child(longVersionNumberHistory).update(constructedData).catch(error => { console.log(error) });
+                }
+            }
+        }
+    }
+    console.log("CVE History tree built");
+}
+
+function writeToDatabaseCVETree(data: any): void {
+    const db = admin.database();
+    const newRef = db.ref('/CVEs');
+    for (const ID in data) {
+        newRef.child(ID).set(data[ID]).catch(error => { console.log(error) });
+    }
+    console.log("CVE tree uploaded");
+}
+
+function pullFromDatabase(version_full: string, version_short: string): void {
+    const db = admin.database();
+    const refCVE = db.ref('/CVEs');
+    let result = null;
+    refCVE.once('value', function (snapshot) {
+        result = snapshot.toJSON();
+        buildBulletinSPLTree(result);
+        buildSPLCVEIDTree(result);
+        buildAOSPVersionASBCVEIDTree(result);
+        buildAOSPVersionCVEIDTree(result);
+        postCVEHistory(JSONData, result, version_short);
+    }).catch(error => { console.log(error) });
+}
+
+function buildBulletinSPLTree(tree: any): void {
+    const db = admin.database();
+    let refBulletinSPL = db.ref('/Bulletin_SPLs');
+    refBulletinSPL.remove().catch(error => { console.log(error) });
```

Comment:
these functions are mixing pure and impure operations. I.e. they have side effects of deleting and updating db elements, in addition to computing a map from a tree. Based on the function name alone, i would have expected it to take a tree and return the map created by lines 159-174